Spark 3.3: Add SparkStagedScan to replace SparkFilesScan

szehon-ho opened the pull request:
This implements the two classes used by RewriteDataFiles for Position Deletes table.  (These existing classes provide short-circuit the read/write of Spark Dataframe, by looking-up a cache of pre-computed ScanTasks).

This is needed for implementing RewriteDeleteFiles, using the same mechanism.

Next step is to implement write of PositionDeletesFile table.  This change is needed first, as I plan to allow write of Position Deletes only if these spark flags are set, as it's meant to be used internally.

szehon-ho commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/AbstractFileRewriteCoordinator.java:
```diff
@@ -0,0 +1,89 @@
+/*
```
This is moved from FileRewriteCoordinator, but made typed.

szehon-ho commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/AbstractScanTaskSetManager.java:
```diff
@@ -0,0 +1,69 @@
+/*
```
This is moved from FileScanTaskSetManager, but made typed

aokolnychyi commented:
Let me take a look at this today/tomorrow.

aokolnychyi commented:
Okay, I went through the change and understand what we are trying to implement.
I'll do a proper review with fresh eyes tomorrow.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/util/TableScanUtil.java:
```diff
@@ -79,6 +79,23 @@ public static CloseableIterable<FileScanTask> splitFiles(
     return CloseableIterable.combine(splitTasks, tasks);
   }
 
+  public static <T extends ScanTask> CloseableIterable<T> splitScanTasks(
```
Instead of doing this, we should offer another method with the following signature.

```
public static <T extends ScanTask> List<ScanTaskGroup<T>> planTaskGroups(
    List<T> tasks,
    long splitSize,
    int lookback,
    long openFileCost) {
  ...
}
```

This will work in the scan for staged tasks as well as for `SparkPartitioningAwareScan`. Under the hood, it will simply delegate to the existing `planTaskGroups` but will wrap list into `CloseableIterable`.

szehon-ho replied:
Done

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/AbstractFileRewriteCoordinator.java:
```diff
@@ -0,0 +1,89 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractFileRewriteCoordinator<F extends ContentFile<F>> {
```
A lot has changed since we added `FileScanTaskSetManager`. I think that was even me? Back then we did not have any other scan task types and our Spark logic wasn't that flexible in terms of supporting those scan task types. That's no longer true.

Instead of offering a file rewrite coordinator for each file type and adapting the existing logic, I'd deprecate existing classes and create a new more flexible path. We will eventually drop the old path.

I'd start by adding a more generic read option.

```
public static final String SCAN_TASK_SET_ID = "scan-task-set-id";
```

We would then create a new task set manager with signature like below (just an idea, can be different).

```
public class ScanTaskSetManager {

  public void stageTasks(Table table, String setId, List<? extends ScanTask> tasks) {
    ...
  }

  @SuppressWarnings("unchecked")
  public <T extends ScanTask> List<T> fetchTasks(Table table, String setId) {
    ...
  }

  @SuppressWarnings("unchecked")
  public <T extends ScanTask> List<T> removeTasks(Table table, String setID) {
    ...
  }
}
```

Then `SparkStagedScanBuilder` and `SparkStagedScan` classes that would support arbitrary scan tasks. The scan implementation will be very straightforward.

```
@Override
protected List<ScanTaskGroup<ScanTask>> taskGroups() {
  if (taskGroups == null) {
    ScanTaskSetManager taskSetManager = ScanTaskSetManager.get();
    List<ScanTask> tasks = taskSetManager.fetchTasks(table(), taskSetId);
    ValidationException.check(
        tasks != null,
        "Task set manager has no tasks for table %s with task set ID %s",
        table().name(),
        taskSetId);

    this.taskGroups = TableScanUtil.planTaskGroups(tasks, splitSize, splitLookback, openFileCost);
  }

  return taskGroups;
}
```

Then new path can be used for data and delete file compaction.

aokolnychyi replied:
This may seem like more work but it will actually make this PR smaller and it is very close to what you did initially.

aokolnychyi replied:
We can keep existing `SparkFilesScanBuilder` and `SparkFilesScan` untouched for now. Just deprecate.

szehon-ho replied:
Done, let me know if its what you mean.

Deprecated SparkFilesScanBuilder and SparkFilesScan.  Can we remove them, and the support for the existing flag "file-scan-task-set-id"?

szehon-ho replied:
(note: I think the comment was for another class only?, so I did not change this class file)

aokolnychyi replied:
My bad. I commented on the wrong class. The rewrite coordinator belongs to writes. Can we include it in the follow-up change for writes? It will be easier to review it there. Let's focus on reading staged delete files in this PR.

szehon-ho replied:
Removed this class

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ScanTaskSetManager.java:
```diff
@@ -0,0 +1,77 @@
+    tasksMap.put(id, tasks);
+  }
+
+  public List<ScanTask> fetchTasks(Table table, String setID) {
```
nit: It was my bad to use all capital letters for `ID` in the original implementation. It is not what we have in other places. Can we replace all usages of `xxxID` to `xxId` in var names and methods?

szehon-ho replied:
Done

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ScanTaskSetManager.java:
```diff
@@ -0,0 +1,77 @@
+    return INSTANCE;
+  }
+
+  public void stageTasks(Table table, String setID, List<ScanTask> tasks) {
```
We should be able to stage any types of tasks. For instance, if I have `List<FileScanTask>`, I should still be able to stage and get these tasks.

I think the signatures should be as follows:

```
public void stageTasks(Table table, String setId, List<? extends ScanTask> tasks) {
  ...
}

@SuppressWarnings("unchecked")
public <T extends ScanTask> List<T> fetchTasks(Table table, String setId) {
  ...
  return (List<T>) tasksMap.get(id);
}
```

We will need to use `List<? extends ScanTask>` in `tasksMap` too.

szehon-ho replied:
Done.

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkReadConf.java:
```diff
@@ -94,6 +94,10 @@ public String fileScanTaskSetId() {
     return confParser.stringConf().option(SparkReadOptions.FILE_SCAN_TASK_SET_ID).parseOptional();
   }
 
+  public String scanTaskSetId() {
```
Shall we also deprecate `fileScanTaskSetId`?

szehon-ho replied:
Done

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkReadOptions.java:
```diff
@@ -65,6 +65,9 @@ private SparkReadOptions() {}
   // Set ID that is used to fetch file scan tasks
   public static final String FILE_SCAN_TASK_SET_ID = "file-scan-task-set-id";
```
Shall we deprecate this property too?

szehon-ho replied:
Done

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkFilesScan.java:
```diff
@@ -32,6 +32,8 @@
 import org.apache.iceberg.util.TableScanUtil;
 import org.apache.spark.sql.SparkSession;
 
+/** @deprecated use {@link SparkStagedScan} instead */
```
Such annotations should also include the version in which we plan to remove it.

```
/** @deprecated will be removed in 1.3.0, use {@link SparkStagedScan} instead */
```

szehon-ho replied:
Done

aokolnychyi replied:
Looks like this was missed.

szehon-ho replied:
Whoops, fixed.

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkStagedScan.java:
```diff
@@ -0,0 +1,114 @@
+  private final String taskSetID;
+  private final long splitSize;
+  private final int splitLookback;
+  private final long splitOpenFileCost;
```
nit: Let's call it `openFileCost` so that our `planTaskGroups` call would fit on one line.

szehon-ho replied:
Done

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkStagedScan.java:
```diff
@@ -0,0 +1,114 @@
+
+class SparkStagedScan extends SparkScan {
+
+  private final String taskSetID;
```
nit: `taskSetID` -> `taskSetId`

szehon-ho replied:
Done

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkStagedScan.java:
```diff
@@ -0,0 +1,114 @@
+    this.splitOpenFileCost = readConf.splitOpenFileCost();
+  }
+
+  protected String taskSetID() {
```
Do we need these methods? We can directly access fields in `equals`.

szehon-ho replied:
Removed

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkStagedScan.java:
```diff
@@ -0,0 +1,114 @@
+  @Override
+  protected List<ScanTaskGroup<ScanTask>> taskGroups() {
+    if (taskGroups == null) {
+
```
nit: Is this new line intentional?

szehon-ho replied:
Removed

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkStagedScan.java:
```diff
@@ -0,0 +1,114 @@
+      List<ScanTask> tasks = taskSetManager.fetchTasks(table(), taskSetID);
+      ValidationException.check(
+          tasks != null,
+          "Task set manager has no tasks for table %s with id %s",
```
nit: `with id ` -> `with task set ID`

szehon-ho replied:
Done

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkStagedScan.java:
```diff
@@ -0,0 +1,114 @@
+  @Override
+  public String toString() {
+    return String.format(
+        "SparkStagedScan(table=%s, type=%s, taskSetID=%s, caseSensitive=%s)",
```
Could you check what naming we use in other scans? Should it be `IcebergStagedScan`?

szehon-ho replied:
Yea it seems to be the general pattern (though SparkCopyOnWriteScan seems exception).  Changed to IcebergStagedScan

aokolnychyi reviewed (commented):
This is getting close. I did a detailed review round and had a few comments. I'd also focus on reading staged scan tasks in this PR and submit rewrite coordinator with PR that adds writes.

szehon-ho commented:
@aokolnychyi thanks for the look.  Addressed the comments.  Also migrated one test of old SparkFilesScan to use SparkStagedScan, to check the API that way.

szehon-ho commented:
I updated the pr title to more reflect the current state of it.

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ScanTaskSetManager.java:
```diff
@@ -0,0 +1,80 @@
+    return INSTANCE;
+  }
+
+  public <T extends ScanTask> void stageTasks(Table table, String setID, List<T> tasks) {
```
nit: `setID` -> `setId`

szehon-ho replied:
Whoops, fixed

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ScanTaskSetManager.java:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  @SuppressWarnings("unchecked")
+  public <T extends ScanTask> List<T> removeTasks(Table table, String setID) {
```
nit: `setID` -> `setId`.

szehon-ho replied:
Whoops, fixed

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkStagedScan.java:
```diff
@@ -0,0 +1,96 @@
+import org.apache.iceberg.util.TableScanUtil;
+import org.apache.spark.sql.SparkSession;
+
+class SparkStagedScan extends SparkScan {
```
Looks good!

aokolnychyi reviewed (approved):
Looks good to me. A few minot nits and good to go.

Let's wait for PR #6965 to be merged first. That one is a big change that is hard to rebase.

szehon-ho commented:
@aokolnychyi thanks, sounds good, will rebase and merge once that is in, then

aokolnychyi commented:
Thanks, @szehon-ho!
